src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,175 @@
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+
+        protected abstract void GetImportedNamespaces(
+            SyntaxNode location,
+            SemanticModel semanticModel,
+            ImmutableHashSet<string>.Builder builder,
+            CancellationToken cancellationToken);
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+            var document = completionContext.Document;
+            var cancellationToken = completionContext.CancellationToken;
+            var workspace = document.Project.Solution.Workspace;
+            var experimentationService = workspace.Services.GetService<IExperimentationService>();
+
+            var importCompletionOptionValue = completionContext.Options.GetOption(CompletionOptions.ShowImportCompletionItems, document.Project.Language);
+
+            // Don't trigger import completion if the option value is "default" and the experiment is disabled for the user. 
+            if (importCompletionOptionValue == false ||
+                (importCompletionOptionValue == null && experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion) != true))
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, completionContext.Position, cancellationToken).ConfigureAwait(false);
+            if (!syntaxContext.IsTypeContext)
+            {
+                return;
+            }
+
+            using (Logger.LogBlock(FunctionId.Completion_TypeImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+            {
+                var project = document.Project;
+                var typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+
+                // Find all namespaces in scope at current cursor location, 
+                // which will be used to filter so the provider only returns out-of-scope types.
+                var namespacesInScope = GetNamespacesInScope(document, syntaxContext, cancellationToken);
+                Action<CompletionItem> handleAccessibleItem = item => AddItems(item, completionContext, namespacesInScope);
+
+                // Get completion items from current project.
+                await typeImportCompletionService.GetAccessibleTopLevelTypesFromProjectAsync(project, handleAccessibleItem, cancellationToken)
+                    .ConfigureAwait(false);
+
+                var compilation = await project.GetCompilationAsync(cancellationToken).ConfigureAwait(false);
+                foreach (var reference in compilation.References)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    var declarationsInReference = ImmutableArray<CompletionItem>.Empty;
+                    if (reference is CompilationReference compilationReference)
```

Comment:
this looks like won't work for cross language p2p reference (c# <-> VB)

you have project so you can just use project.ProjectReferences rather than go through compilation to get back to roslyn solution. you only need to do that when you don't have roslyn solutin. in this case, you already have. so, you can just use that.

Reply:
Yea, I have already switched to using Project.ProjectReferences and PRoject.MetadataRefrences instead